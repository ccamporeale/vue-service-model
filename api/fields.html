<!DOCTYPE html>
<html lang="en-US">
  <head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width,initial-scale=1">
    <title>Fields API | vue-service-model</title>
    <meta name="generator" content="VuePress 1.5.0">
    <link rel="icon" href="/vue-service-model/logo.png">
    <meta name="description" content="Vue.js library for handling REST service requests with caching, aggregation and model definitions">
    <link rel="preload" href="/vue-service-model/assets/css/0.styles.a218223d.css" as="style"><link rel="preload" href="/vue-service-model/assets/js/app.c2b3a9a1.js" as="script"><link rel="preload" href="/vue-service-model/assets/js/3.45671bfb.js" as="script"><link rel="preload" href="/vue-service-model/assets/js/8.64e7feff.js" as="script"><link rel="prefetch" href="/vue-service-model/assets/js/10.22689168.js"><link rel="prefetch" href="/vue-service-model/assets/js/11.ba383af0.js"><link rel="prefetch" href="/vue-service-model/assets/js/12.375d0476.js"><link rel="prefetch" href="/vue-service-model/assets/js/13.804a0c41.js"><link rel="prefetch" href="/vue-service-model/assets/js/14.d6887d0c.js"><link rel="prefetch" href="/vue-service-model/assets/js/15.f93a33c2.js"><link rel="prefetch" href="/vue-service-model/assets/js/16.568206dd.js"><link rel="prefetch" href="/vue-service-model/assets/js/17.c3d9d7ac.js"><link rel="prefetch" href="/vue-service-model/assets/js/18.247cde34.js"><link rel="prefetch" href="/vue-service-model/assets/js/19.1d1f842e.js"><link rel="prefetch" href="/vue-service-model/assets/js/20.1cb7654d.js"><link rel="prefetch" href="/vue-service-model/assets/js/21.7db86893.js"><link rel="prefetch" href="/vue-service-model/assets/js/22.e933c31d.js"><link rel="prefetch" href="/vue-service-model/assets/js/4.973410df.js"><link rel="prefetch" href="/vue-service-model/assets/js/5.c2a7c680.js"><link rel="prefetch" href="/vue-service-model/assets/js/6.b5c2df69.js"><link rel="prefetch" href="/vue-service-model/assets/js/7.612f3f77.js"><link rel="prefetch" href="/vue-service-model/assets/js/9.4a145423.js"><link rel="prefetch" href="/vue-service-model/assets/js/vendors~docsearch.a7fb83ab.js">
    <link rel="stylesheet" href="/vue-service-model/assets/css/0.styles.a218223d.css">
  </head>
  <body>
    <div id="app" data-server-rendered="true"><div class="theme-container"><header class="navbar"><div class="sidebar-button"><svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" role="img" viewBox="0 0 448 512" class="icon"><path fill="currentColor" d="M436 124H12c-6.627 0-12-5.373-12-12V80c0-6.627 5.373-12 12-12h424c6.627 0 12 5.373 12 12v32c0 6.627-5.373 12-12 12zm0 160H12c-6.627 0-12-5.373-12-12v-32c0-6.627 5.373-12 12-12h424c6.627 0 12 5.373 12 12v32c0 6.627-5.373 12-12 12zm0 160H12c-6.627 0-12-5.373-12-12v-32c0-6.627 5.373-12 12-12h424c6.627 0 12 5.373 12 12v32c0 6.627-5.373 12-12 12z"></path></svg></div> <a href="/vue-service-model/" class="home-link router-link-active"><img src="/vue-service-model/logo.png" alt="vue-service-model" class="logo"> <span class="site-name can-hide">vue-service-model</span></a> <div class="links"><form id="search-form" role="search" class="algolia-search-wrapper search-box"><input id="algolia-search-input" class="search-query"></form> <nav class="nav-links can-hide"><div class="nav-item"><a href="/vue-service-model/guide/" class="nav-link">
  Guide
</a></div><div class="nav-item"><a href="/vue-service-model/api/" class="nav-link router-link-active">
  API Reference
</a></div><div class="nav-item"><a href="https://github.com/freakzlike/vue-service-model/releases" target="_blank" rel="noopener noreferrer" class="nav-link external">
  Release Notes
  <svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg></a></div> <a href="https://github.com/freakzlike/vue-service-model" target="_blank" rel="noopener noreferrer" class="repo-link">
    GitHub
    <svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg></a></nav></div></header> <div class="sidebar-mask"></div> <aside class="sidebar"><nav class="nav-links"><div class="nav-item"><a href="/vue-service-model/guide/" class="nav-link">
  Guide
</a></div><div class="nav-item"><a href="/vue-service-model/api/" class="nav-link router-link-active">
  API Reference
</a></div><div class="nav-item"><a href="https://github.com/freakzlike/vue-service-model/releases" target="_blank" rel="noopener noreferrer" class="nav-link external">
  Release Notes
  <svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg></a></div> <a href="https://github.com/freakzlike/vue-service-model" target="_blank" rel="noopener noreferrer" class="repo-link">
    GitHub
    <svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg></a></nav>  <ul class="sidebar-links"><li><a href="/vue-service-model/" class="sidebar-link">Home</a></li><li><section class="sidebar-group depth-0"><a href="/vue-service-model/guide/" class="sidebar-heading clickable"><span>Guide</span> <!----></a> <ul class="sidebar-links sidebar-group-items"><li><a href="/vue-service-model/guide/getting-started.html" class="sidebar-link">Getting started</a></li><li><a href="/vue-service-model/guide/installation.html" class="sidebar-link">Installation</a></li><li><a href="/vue-service-model/guide/models.html" class="sidebar-link">Models</a></li><li><a href="/vue-service-model/guide/fields.html" class="sidebar-link">Fields</a></li><li><a href="/vue-service-model/guide/field-types.html" class="sidebar-link">Field types</a></li><li><a href="/vue-service-model/guide/service-model.html" class="sidebar-link">ServiceModel</a></li><li><a href="/vue-service-model/guide/model-manager.html" class="sidebar-link">ModelManager</a></li><li><a href="/vue-service-model/guide/components.html" class="sidebar-link">Components</a></li><li><a href="/vue-service-model/guide/configuration.html" class="sidebar-link">Configuration</a></li></ul></section></li><li><section class="sidebar-group depth-0"><p class="sidebar-heading open"><span>API</span> <!----></p> <ul class="sidebar-links sidebar-group-items"><li><a href="/vue-service-model/api/base-model.html" class="sidebar-link">BaseModel API</a></li><li><a href="/vue-service-model/api/service-model.html" class="sidebar-link">ServiceModel API</a></li><li><a href="/vue-service-model/api/fields.html" class="active sidebar-link">Fields API</a></li><li><a href="/vue-service-model/api/model-manager.html" class="sidebar-link">ModelManager API</a></li><li><a href="/vue-service-model/api/configuration.html" class="sidebar-link">Configuration API</a></li></ul></section></li></ul> </aside> <main class="page"> <div class="theme-default-content content__default"><h1 id="fields-api"><a href="#fields-api" class="header-anchor">#</a> Fields API</h1> <div class="language-typescript extra-class"><pre class="language-typescript"><code><span class="token keyword">class</span> <span class="token class-name">Field</span> <span class="token punctuation">{</span>
  <span class="token comment">// Constructor takes field definition</span>
  <span class="token keyword">constructor</span> <span class="token punctuation">(</span>def<span class="token operator">:</span> FieldDef<span class="token punctuation">)</span>

  <span class="token comment">// Clone field instance</span>
  <span class="token keyword">public</span> <span class="token function">clone</span> <span class="token punctuation">(</span><span class="token punctuation">)</span><span class="token operator">:</span> Field

  <span class="token comment">// Returns field name (Which has been set as key at fieldsDef.</span>
  <span class="token comment">// Will throw FieldNotBoundException in case field has not been bound to a model</span>
  <span class="token keyword">public</span> <span class="token keyword">get</span> <span class="token function">name</span> <span class="token punctuation">(</span><span class="token punctuation">)</span><span class="token operator">:</span> <span class="token builtin">string</span>

  <span class="token comment">// Returns either attributeName from fieldsDef or field name</span>
  <span class="token keyword">public</span> <span class="token keyword">get</span> <span class="token function">attributeName</span> <span class="token punctuation">(</span><span class="token punctuation">)</span><span class="token operator">:</span> <span class="token builtin">string</span>

  <span class="token comment">// Returns field definition</span>
  <span class="token keyword">public</span> <span class="token keyword">get</span> <span class="token function">definition</span> <span class="token punctuation">(</span><span class="token punctuation">)</span><span class="token operator">:</span> FieldDef

  <span class="token comment">// Assigned model</span>
  <span class="token comment">// Will throw FieldNotBoundException in case field has not been bound to a model</span>
  <span class="token keyword">public</span> <span class="token keyword">get</span> <span class="token function">model</span> <span class="token punctuation">(</span><span class="token punctuation">)</span><span class="token operator">:</span> BaseModel

  <span class="token comment">// Returns async field value from data by calling valueGetter with data of assigned model</span>
  <span class="token keyword">public</span> <span class="token keyword">get</span> <span class="token function">value</span> <span class="token punctuation">(</span><span class="token punctuation">)</span><span class="token operator">:</span> <span class="token builtin">any</span>

  <span class="token comment">// Sets field value to model data by calling valueSetter with data of assigned model</span>
  <span class="token keyword">public</span> <span class="token keyword">set</span> <span class="token function">value</span> <span class="token punctuation">(</span>value<span class="token operator">:</span> <span class="token builtin">any</span><span class="token punctuation">)</span>

  <span class="token comment">// Returns async field label from field definition</span>
  <span class="token keyword">public</span> <span class="token keyword">get</span> <span class="token function">label</span> <span class="token punctuation">(</span><span class="token punctuation">)</span><span class="token operator">:</span> <span class="token builtin">Promise</span><span class="token operator">&lt;</span><span class="token builtin">string</span><span class="token operator">&gt;</span>

  <span class="token comment">// Returns async field hint from field definition</span>
  <span class="token keyword">public</span> <span class="token keyword">get</span> <span class="token function">hint</span> <span class="token punctuation">(</span><span class="token punctuation">)</span><span class="token operator">:</span> <span class="token builtin">Promise</span><span class="token operator">&lt;</span><span class="token builtin">string</span><span class="token operator">&gt;</span>

  <span class="token comment">// Returns boolean whether field is a primary key</span>
  <span class="token keyword">public</span> <span class="token keyword">get</span> <span class="token function">isPrimaryKey</span> <span class="token punctuation">(</span><span class="token punctuation">)</span><span class="token operator">:</span> <span class="token builtin">boolean</span>

  <span class="token comment">// Retrieve value from data structure according to attributeName</span>
  <span class="token comment">// Uses nested syntax from attributeName (e.g. &quot;address.city&quot; -&gt; {address: {city: 'New York'}})</span>
  <span class="token comment">// Will return null if value is not available</span>
  <span class="token keyword">public</span> <span class="token function">valueGetter</span> <span class="token punctuation">(</span>data<span class="token operator">:</span> <span class="token builtin">any</span><span class="token punctuation">)</span><span class="token operator">:</span> <span class="token builtin">any</span>

  <span class="token comment">// Set value to data by using attributeName</span>
  <span class="token comment">// Will create nested structure from attributeName (e.g. &quot;address.city&quot; -&gt; {address: {city: 'New York'}})</span>
  <span class="token keyword">public</span> <span class="token function">valueSetter</span> <span class="token punctuation">(</span>value<span class="token operator">:</span> <span class="token builtin">any</span><span class="token punctuation">,</span> data<span class="token operator">:</span> Dictionary<span class="token operator">&lt;</span><span class="token builtin">any</span><span class="token operator">&gt;</span><span class="token punctuation">)</span><span class="token operator">:</span> <span class="token keyword">void</span>

  <span class="token comment">// Display component to render when displaying value with &lt;display-field/&gt;</span>
  <span class="token comment">// For more information see Field - Rendering </span>
  <span class="token keyword">public</span> <span class="token keyword">get</span> <span class="token function">displayComponent</span> <span class="token punctuation">(</span><span class="token punctuation">)</span><span class="token operator">:</span> <span class="token builtin">Promise</span><span class="token operator">&lt;</span>ComponentModule<span class="token operator">&gt;</span>

  <span class="token comment">// Async function to prepare before displayRender gets called</span>
  <span class="token comment">// Can return any data which needs to be resolved for displayRender</span>
  <span class="token keyword">public</span> <span class="token keyword">async</span> <span class="token function">prepareDisplayRender</span> <span class="token punctuation">(</span><span class="token punctuation">)</span><span class="token operator">:</span> <span class="token builtin">Promise</span><span class="token operator">&lt;</span><span class="token builtin">any</span><span class="token operator">&gt;</span>

  <span class="token comment">// Simple Vue render function when using default displayComponent when displaying value with &lt;display-field/&gt;</span>
  <span class="token comment">// For more information see Field - Rendering </span>
  <span class="token keyword">public</span> <span class="token function">displayRender</span> <span class="token punctuation">(</span>h<span class="token operator">:</span> CreateElement<span class="token punctuation">,</span> renderData<span class="token operator">:</span> <span class="token builtin">any</span><span class="token punctuation">)</span><span class="token operator">:</span> VNode

  <span class="token comment">// Input component to render when showing input for field with &lt;input-field/&gt;</span>
  <span class="token comment">// For more information see Field - Rendering </span>
  <span class="token keyword">public</span> <span class="token keyword">get</span> <span class="token function">inputComponent</span> <span class="token punctuation">(</span><span class="token punctuation">)</span><span class="token operator">:</span> <span class="token builtin">Promise</span><span class="token operator">&lt;</span>ComponentModule<span class="token operator">&gt;</span>

  <span class="token comment">// Async function to prepare before inputRender gets called</span>
  <span class="token comment">// Can return any data which needs to be resolved for inputRender</span>
  <span class="token keyword">public</span> <span class="token keyword">async</span> <span class="token function">prepareInputRender</span> <span class="token punctuation">(</span><span class="token punctuation">)</span><span class="token operator">:</span> <span class="token builtin">Promise</span><span class="token operator">&lt;</span><span class="token builtin">any</span><span class="token operator">&gt;</span>

  <span class="token comment">// Simple Vue render function when using default inputComponent for input of field value with &lt;input-field/&gt;</span>
  <span class="token comment">// For more information see Field - Rendering </span>
  <span class="token keyword">public</span> <span class="token function">inputRender</span> <span class="token punctuation">(</span>h<span class="token operator">:</span> CreateElement<span class="token punctuation">,</span> renderData<span class="token operator">:</span> <span class="token builtin">any</span><span class="token punctuation">)</span><span class="token operator">:</span> VNode

<span class="token punctuation">}</span>
</code></pre></div></div> <footer class="page-edit"><div class="edit-link"><a href="https://github.com/freakzlike/vue-service-model/edit/master/docs/api/fields.md" target="_blank" rel="noopener noreferrer">Edit this page on GitHub</a> <svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg></div> <div class="last-updated"><span class="prefix">Last Updated:</span> <span class="time">5/3/2020, 8:56:57 AM</span></div></footer> <div class="page-nav"><p class="inner"><span class="prev">
      ←
      <a href="/vue-service-model/api/service-model.html" class="prev">
        ServiceModel API
      </a></span> <span class="next"><a href="/vue-service-model/api/model-manager.html">
        ModelManager API
      </a>
      →
    </span></p></div> </main></div><div class="global-ui"></div></div>
    <script src="/vue-service-model/assets/js/app.c2b3a9a1.js" defer></script><script src="/vue-service-model/assets/js/3.45671bfb.js" defer></script><script src="/vue-service-model/assets/js/8.64e7feff.js" defer></script>
  </body>
</html>
