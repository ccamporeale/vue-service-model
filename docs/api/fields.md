# Fields API

```typescript
class Field {
  // Constructor takes field definition
  constructor (def: FieldDef)

  // Clone field instance
  public clone (): Field

  // Returns field name (Which has been set as key at fieldsDef.
  // Will throw FieldNotBoundException in case field has not been bound to a model
  public get name (): string

  // Returns either attributeName from fieldsDef or field name
  public get attributeName (): string

  // Returns field definition
  public get definition (): FieldDef

  // Assigned model
  // Will throw FieldNotBoundException in case field has not been bound to a model
  public get model (): BaseModel

  // Returns async field value from data by calling valueGetter with data of assigned model
  public get value (): any

  // Sets field value to model data by calling valueSetter with data of assigned model
  public set value (value: any)

  // Returns async field label from field definition
  public get label (): Promise<string>

  // Returns async field hint from field definition
  public get hint (): Promise<string>

  // Returns async field options with validation and default values depending on field type
  public get options (): Promise<FieldTypeOptions>

  // Validate field options and set default values depending on field type
  protected async validateOptions (options: FieldTypeOptions): Promise<FieldTypeOptions>

  // Returns boolean whether field is a primary key
  public get isPrimaryKey (): boolean

  // Retrieve value from data structure according to attributeName
  // Uses nested syntax from attributeName (e.g. "address.city" -> {address: {city: 'New York'}})
  // Will return null if value is not available
  public valueGetter (data: any): any

  // Set value to data by using attributeName
  // Will create nested structure from attributeName (e.g. "address.city" -> {address: {city: 'New York'}})
  public valueSetter (value: any, data: Dictionary<any>): void

  // Display component to render when displaying value with <display-field/>
  // For more information see Field - Rendering 
  public get displayComponent (): Promise<ComponentModule>

  // Async function to prepare before displayRender gets called
  // Can return any data which needs to be resolved for displayRender
  public async prepareDisplayRender (): Promise<any>

  // Simple Vue render function when using default displayComponent when displaying value with <display-field/>
  // For more information see Field - Rendering 
  public displayRender (h: CreateElement, renderData: any): VNode

  // Input component to render when showing input for field with <input-field/>
  // For more information see Field - Rendering 
  public get inputComponent (): Promise<ComponentModule>

  // Async function to prepare before inputRender gets called
  // Can return any data which needs to be resolved for inputRender
  public async prepareInputRender (): Promise<any>

  // Simple Vue render function when using default inputComponent for input of field value with <input-field/>
  // For more information see Field - Rendering 
  public inputRender (h: CreateElement, renderData: any): VNode

}
```
