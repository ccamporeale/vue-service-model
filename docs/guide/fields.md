# Fields

[[toc]]

## Field definition

When instantiating a new field you can provide a specific definition for the field.

```js
class Album extends BaseModel {
  [...]

  static fieldsDef = {
    id: new Field({primaryKey: true}),
    title: new Field({label: 'Album title'})
  }
}
```

Field definition structure:
```js
{
  // String which key should be used to retrieve value from. See Attribute name for more information
  // Optional: default uses key from fieldsDef 
  attributeName: 'title',

  // Label of field. See Field label and hint for more information
  // Optional: Can either be a string, function or promise
  label: () => Promise.resolve('Title'),

  // Hint of field. See Field label and hint for more information
  // Optional: Can either be a string, function or promise
  hint: 'Title of album',

  // Boolean flag whether field is a primary key
  // Optional: default is false
  primaryKey: true
}
```

### Attribute name (`attributeName`)

By default the key of your field in your `fieldsDef` (e.g. `first_name`) will be used to retrieve the value from the model data.
You can also set the `attributeName` when instantiating the field. It is also possible to access nested data when using dot-notation in `attributeName`.
If you need a more specific way to retrieve the value of a field from your data then have a look at [Custom/Computed fields](#customcomputed-fields).

```js
class MyModel extends BaseModel {
  [...]

  static fieldsDef = {
    name: new Field({attributeName: 'username'}),
    address_city: new Field({attributeName: 'address.city'}),
    address_street: new Field({attributeName: 'address.street.name'})
  }
}

const myObj = new MyModel({
  username: 'joe_bloggs',
  address: {
    city: 'New York',
    street: {
      name: 'Fifth Avenue'
    } 
  }
})

await myObj.val.name // output: joe_bloggs
await myObj.val.address_city // output: New York
await myObj.val.address_street // output: Fifth Avenue
```

### Field label and hint (`label`, `hint`)

With the `label` property you can set a descriptive name for your field. `hint` is used to provide a detail description of your field. `label` and `hint` can either be a `string` or a `function` which should return a `string` or a `Promise`.
You can access your label/hint with the `label`/`hint` property of your field instance which will always return a `Promise`.

```js
class MyModel extends BaseModel {
  [...]

  static fieldsDef = {
    first_name: new Field({
      label: 'First name',
      hint: () => 'First name of the employee'
    })
  }
}

[...]

const firstNameField = myObj.getField('first_name')

await firstNameField.label // output: First name
await firstNameField.hint // output: First name of the employee
```

## Field API

```typescript
class Field {
  // Constructor takes field definition
  constructor (def: FieldDef)

  // Clone field instance
  public clone (): Field

  // Returns field name (Which has been set as key at fieldsDef.
  // Will throw FieldNotBoundException in case field has not been bound to a model
  public get name (): string

  // Returns either attributeName from fieldsDef or field name
  public get attributeName (): string

  // Returns field definition
  public get definition (): FieldDef

  // Assigned model
  // Will throw FieldNotBoundException in case field has not been bound to a model
  public get model (): BaseModel

  // Returns async field value from data by calling valueGetter with data of assigned model
  public get value (): any

  // Sets field value to model data by calling valueSetter with data of assigned model
  public set value (value: any)

  // Returns async field label from field definition
  public get label (): Promise<string>

  // Returns async field hint from field definition
  public get hint (): Promise<string>

  // Returns boolean whether field is a primary key
  public get isPrimaryKey (): boolean

  // Retrieve value from data structure according to attributeName
  // Uses nested syntax from attributeName (e.g. "address.city" -> {address: {city: 'New York'}})
  // Will return null if value is not available
  public valueGetter (data: any): any

  // Set value to data by using attributeName
  // Will create nested structure from attributeName (e.g. "address.city" -> {address: {city: 'New York'}})
  public valueSetter (value: any, data: Dictionary<any>): void

  // Display component to render when displaying value with <display-field/>
  // For more information see Field - Rendering 
  public get displayComponent (): Promise<ComponentModule>

  // Simple Vue render function when using default displayComponent when displaying value with <display-field/>
  // For more information see Field - Rendering 
  public displayRender (h: CreateElement, resolvedValue: any): VNode

  // Input component to render when showing input for field with <input-field/>
  // For more information see Field - Rendering 
  public get inputComponent (): Promise<ComponentModule>

  // Simple Vue render function when using default inputComponent for input of field value with <input-field/>
  // For more information see Field - Rendering 
  public inputRender (h: CreateElement, resolvedValue: any): VNode

}
```

## Custom/Computed fields

In case you want to define your own field class you just need to extend from `Field`. By overwriting the `valueGetter` method you are able to map the field value by yourself and create computed values.

```js
class FullNameField extends Field {
  valueGetter (data) {
    return data ? data.first_name + ' ' + data.last_name : null
  }
}

class MyModel extends BaseModel {
  [...]

  static fieldsDef = {
    full_name: new FullNameField()
  }
}

const myObj = new MyModel({
  first_name: 'Joe',
  last_name: 'Bloggs'
})

await myObj.val.full_name // output: Joe Bloggs
```

## Field types

Different field types will be added with future releases.

## Rendering

By using the [`DisplayField`](#displayfield) component you can render the value of a field for displaying purpose
and [`InputField`](#inputfield) when you want to render the input component for this field.
To customize the different output which should be rendered you can either set a `displayRender`/`inputRender` or a custom `displayComponent`/`inputComponent`.

The `displayRender`/`inputRender` can be used for small changes of the output and is a simple render function (See [Vue.js render function](https://vuejs.org/v2/guide/render-function.html)).

```js
class RedTextField extends Field {
  // Render a red text for display
  displayRender (h, resolvedValue) {
    return h('span', {
      style: {
        color: 'red'
      }
    }, resolvedValue)
  }

  // Render an text input field for input
  inputRender (h, resolvedValue) {
    return h('input', {
      attrs: {
        type: 'text',
        value: resolvedValue
      },
      on: {
        input: (event) => {
          // Set input value to field.value
          this.value = event.target.value
        }
      }
    })
  }
}
```

In case you need to do more specific rendering you can also set your own component which will be rendered when using [`DisplayField`](/guide/components.html#displayfield) on your custom field.
If you want to change the input component you can extend from `BaseInputFieldRender` and overwrite the `inputComponent` method of your field.
For easy usage with async values if recommend you to take a look at [vue-async-computed](https://github.com/foxbenjaminfox/vue-async-computed).

```js
// CustomField.js
class CustomField extends Field {
  get displayComponent () {
    return import('./CustomFieldComponent')
  }
}

// CustomFieldComponent.vue
<template>
  <span>{{ fieldValue }}</span>
</template>

<script>
  import {BaseDisplayFieldRender} from 'vue-service-model'

  export default {
    extends: BaseDisplayFieldRender,
    asyncComputed: {
      fieldValue () {
        return this.field.value      
      }  
    }
  }
</script>
```