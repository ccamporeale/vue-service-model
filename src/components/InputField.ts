import { CreateElement, VNode, Component } from 'vue'
import mixins from '../utils/mixins'
import FieldPropsMixin from './FieldPropsMixin'
import LoadingSlotMixin from './LoadingSlotMixin'
import { configHandler } from '../utils/ConfigHandler'

interface ComponentData {
  inputComponent: Component | null
}

/**
 * Main component to display value of a field
 */
export default mixins(LoadingSlotMixin, FieldPropsMixin).extend({
  name: 'InputField',
  inheritAttrs: false,

  data: (): ComponentData => ({
    inputComponent: null
  }),

  asyncComputed: {
    async inputComponent (): Promise<Component | null> {
      configHandler.checkWarningUseAsyncComputed()
      return this.resolveInputComponent()
    }
  },

  watch: {
    fieldObj () {
      this.setResolveInputComponent()
    }
  },

  created () {
    this.setResolveInputComponent()
  },

  methods: {
    async setResolveInputComponent () {
      if (!configHandler.useAsyncComputed()) {
        this.inputComponent = await this.resolveInputComponent()
      }
    },

    async resolveInputComponent () {
      if (this.fieldObj) {
        const componentModule = await this.fieldObj.inputComponent
        return componentModule.default
      } else {
        return null
      }
    }
  },

  render (h: CreateElement): VNode {
    if (this.inputComponent && this.fieldObj) {
      return h(this.inputComponent, {
        props: {
          field: this.fieldObj
        }
      })
    } else {
      return this.renderLoading(h)
    }
  }
})
