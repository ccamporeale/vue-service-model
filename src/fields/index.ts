export { Field } from './Field'
export { FormatStringField } from './FormatStringField'

export { CharField } from './CharField'
export { BooleanField } from './BooleanField'
export { IntegerField } from './IntegerField'
export { DecimalField } from './DecimalField'
