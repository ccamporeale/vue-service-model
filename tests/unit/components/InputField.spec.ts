import Vue, { mount, Wrapper } from '@vue/test-utils'
import { Field } from '@/fields/Field'
import InputField from '@/components/InputField'
import { BaseModel } from '@/models/BaseModel'

describe('components/InputField', () => {
  const modelData = {
    name: 'Name 1',
    description: 'Description'
  }

  class TestModel extends BaseModel {
    static fieldsDef = {
      name: new Field(),
      description: new Field()
    }
  }

  const model = new TestModel(modelData)

  const renderInputField = async (wrapper: Wrapper<Vue>) => {
    await wrapper.vm.$nextTick()
    await wrapper.vm.$nextTick()
  }

  it('should render correctly with model and fieldName', async () => {
    const wrapper = mount(InputField, {
      propsData: {
        model: model,
        fieldName: 'name'
      }
    })
    expect(wrapper.vm.inputComponent).toBeNull()
    expect(wrapper.html()).toBeUndefined()

    await renderInputField(wrapper)

    expect(wrapper.vm.inputComponent).not.toBeNull()
    await wrapper.vm.$nextTick()

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render correctly with field', async () => {
    const wrapper = mount(InputField, {
      propsData: {
        field: model.getField('name')
      }
    })
    expect(wrapper.vm.inputComponent).toBeNull()
    expect(wrapper.html()).toBeUndefined()

    await renderInputField(wrapper)

    expect(wrapper.vm.inputComponent).not.toBeNull()
    await wrapper.vm.$nextTick()

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should not render without model', async () => {
    const wrapper = mount(InputField, {
      propsData: {
        model: null,
        fieldName: 'name'
      }
    })

    await renderInputField(wrapper)

    expect(wrapper.vm.inputComponent).toBeNull()
    await wrapper.vm.$nextTick()
    expect(wrapper.html()).toBeUndefined()
  })

  it('should not render when model reset', async () => {
    const wrapper = mount(InputField, {
      propsData: {
        model: model,
        fieldName: 'name'
      }
    })

    await renderInputField(wrapper)

    expect(wrapper.vm.inputComponent).not.toBeNull()
    await wrapper.vm.$nextTick()
    const inputElement = wrapper.find('input')
    expect(inputElement.is('input')).toBe(true)
    expect(inputElement.attributes('value')).toBe(modelData.name)

    wrapper.setProps({ model: null })

    expect(wrapper.html()).toBeUndefined()

    await wrapper.vm.$nextTick()
    expect(wrapper.vm.inputComponent).toBeNull()
  })

  it('should render correct when field name changed', async () => {
    const wrapper = mount(InputField, {
      propsData: {
        model: model,
        fieldName: 'name'
      }
    })

    await renderInputField(wrapper)

    expect(wrapper.vm.inputComponent).not.toBeNull()
    await wrapper.vm.$nextTick()

    const inputElement = wrapper.find('input')
    expect(inputElement.is('input')).toBe(true)
    expect(inputElement.attributes('value')).toBe(modelData.name)

    wrapper.setProps({ fieldName: 'description' })

    await renderInputField(wrapper)
    expect(inputElement.attributes('value')).toBe(modelData.description)
  })

  it('should render correct when field changed', async () => {
    const wrapper = mount(InputField, {
      propsData: {
        field: model.getField('name')
      }
    })

    await renderInputField(wrapper)

    expect(wrapper.vm.inputComponent).not.toBeNull()
    await wrapper.vm.$nextTick()

    const inputElement = wrapper.find('input')
    expect(inputElement.is('input')).toBe(true)
    expect(inputElement.attributes('value')).toBe(modelData.name)

    wrapper.setProps({ field: model.getField('description') })

    await renderInputField(wrapper)
    expect(inputElement.attributes('value')).toBe(modelData.description)
  })
})
