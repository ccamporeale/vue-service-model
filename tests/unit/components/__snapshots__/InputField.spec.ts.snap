// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/InputField should render correctly with field 1`] = `<input type="text" value="Name 1">`;

exports[`components/InputField should render correctly with model and fieldName 1`] = `<input type="text" value="Name 1">`;
